File: lingting-aws/src/main/java/live/lingting/framework/aws/AwsS3Client.java
package live.lingting.framework.aws;

import live.lingting.framework.aws.exception.AwsS3Exception;
import live.lingting.framework.aws.policy.Acl;
import live.lingting.framework.aws.s3.AwsS3Properties;
import live.lingting.framework.aws.s3.AwsS3Request;
import live.lingting.framework.aws.s3.AwsS3SingV4;
import live.lingting.framework.aws.s3.AwsS3Utils;
import live.lingting.framework.http.HttpResponse;
import live.lingting.framework.http.api.ApiClient;
import live.lingting.framework.http.header.HttpHeaders;
import live.lingting.framework.util.StringUtils;
import live.lingting.framework.value.multi.StringMultiValue;
import lombok.Setter;

import java.net.http.HttpRequest;
import java.time.LocalDateTime;
import java.util.function.BiConsumer;

import static live.lingting.framework.aws.s3.AwsS3SingV4.DATETIME_FORMATTER;
import static live.lingting.framework.aws.s3.AwsS3Utils.HEADER_ACL;
import static live.lingting.framework.aws.s3.AwsS3Utils.HEADER_CONTENT_SHA256;
import static live.lingting.framework.aws.s3.AwsS3Utils.HEADER_DATE;
import static live.lingting.framework.aws.s3.AwsS3Utils.HEADER_TOKEN;
import static live.lingting.framework.aws.s3.AwsS3Utils.PAYLOAD_UNSIGNED;

/**
 * @author lingting 2024-09-19 15:02
 */
public abstract class AwsS3Client extends ApiClient<AwsS3Request> {

	protected final AwsS3Properties properties;

	protected final String ak;

	protected final String sk;

	protected final String token;

	protected final Acl acl;

	protected final String bucket;

	@Setter
	protected BiConsumer<AwsS3Request, HttpResponse> onFailed = (awsS3Request, response) -> {
		String string = response.string();
		log.error("AwsS3 call error! uri: {}; code: {}; body:\n{}", response.uri(), response.code(), string);
		throw new AwsS3Exception("request error! code: " + response.code());
	};

	protected AwsS3Client(AwsS3Properties properties) {
		super(properties.host());
		this.properties = properties;
		this.ak = properties.getAk();
		this.sk = properties.getSk();
		this.token = properties.getToken();
		this.acl = properties.getAcl();
		this.bucket = properties.getBucket();
	}

	@Override
	protected void customize(AwsS3Request request, HttpHeaders headers, HttpRequest.BodyPublisher publisher,
			StringMultiValue params) {
		if (request.getAcl() != null) {
			headers.put(HEADER_ACL, request.getAcl().getValue());
		}

		LocalDateTime now = LocalDateTime.now();
		String date = AwsS3Utils.format(now, DATETIME_FORMATTER);
		headers.put(HEADER_DATE, date);
		headers.put(HEADER_CONTENT_SHA256, PAYLOAD_UNSIGNED);

		if (StringUtils.hasText(token)) {
			headers.put(HEADER_TOKEN, token);
		}
		AwsS3SingV4 sing = AwsS3SingV4.builder()
			.dateTime(now)
			.method(request.method())
			.path(request.path())
			.headers(headers)
			.bodySha256(PAYLOAD_UNSIGNED)
			.params(params)
			.region(properties.getRegion())
			.ak(ak)
			.sk(sk)
			.bucket(bucket)
			.build();

		String authorization = sing.calculate();
		headers.authorization(authorization);
	}

	@Override
	protected HttpResponse checkout(AwsS3Request request, HttpResponse response) {
		if (!response.is2xx()) {
			onFailed.accept(request, response);
		}
		return response;
	}

}

File: lingting-aws/src/main/kotlin/live/lingting/framework/aws/AwsS3Client.kt
package live.lingting.framework.aws


import live.lingting.framework.aws.policy.Acl
import live.lingting.framework.aws.properties.S3Properties
import live.lingting.framework.aws.s3.AwsS3PreRequest
import live.lingting.framework.aws.s3.AwsS3Request
import live.lingting.framework.aws.s3.impl.AwsS3DefaultListener
import live.lingting.framework.aws.s3.interfaces.AwsS3Listener
import live.lingting.framework.aws.s3.response.AwsS3PreSignedResponse
import live.lingting.framework.http.HttpResponse
import live.lingting.framework.http.HttpUrlBuilder
import live.lingting.framework.http.api.ApiClient
import live.lingting.framework.http.body.Body
import live.lingting.framework.http.header.HttpHeaders
import live.lingting.framework.jackson.JacksonUtils
import live.lingting.framework.time.DateTime
import java.io.ByteArrayInputStream
import java.nio.charset.Charset
import java.nio.charset.StandardCharsets

/**
 * @author lingting 2024-09-19 15:02
 */
abstract class AwsS3Client protected constructor(val properties: S3Properties) :
    ApiClient<AwsS3Request>(properties.host(), properties.ssl) {

    open val charset: Charset = StandardCharsets.UTF_8

    val ak: String = properties.ak

    val sk: String = properties.sk

    val token: String? = properties.token

    val acl: Acl? = properties.acl

    val bucket: String = properties.bucket

    var listener: AwsS3Listener = AwsS3DefaultListener(this)

    override fun urlBuilder(): HttpUrlBuilder {
        return properties.urlBuilder()
    }

    override fun checkout(request: AwsS3Request, response: HttpResponse): HttpResponse {
        if (!response.is2xx) {
            listener.onFailed(request, response)
        }
        return response
    }

    override fun call(
        urlBuilder: HttpUrlBuilder,
        r: AwsS3Request,
        headers: HttpHeaders,
        body: Body
    ): HttpResponse {
        if (r.acl != null) {
            headers.put(AwsUtils.HEADER_ACL, r.acl!!.value)
        }

        if (!token.isNullOrBlank()) {
            headers.put(AwsUtils.HEADER_TOKEN, token)
        }

        r.meta.forEach { k, vs ->
            val key = if (k.startsWith(AwsUtils.HEADER_PREFIX_META)) k else "${AwsUtils.HEADER_PREFIX_META}$k"
            headers.addAll(key, vs)
        }

        val signer = listener.onSign(r, headers, urlBuilder)

        val current = DateTime.current()

        val signed =
            if (r is AwsS3PreRequest) signer.signed(current, r.expire)
            else signer.signed(current)

        signed.fill(headers, urlBuilder)
        if (r !is AwsS3PreRequest) {
            val request = buildRequest(urlBuilder, headers, r, body)
            return call(r, request)
        }

        val url = urlBuilder.build()
        val value = AwsS3PreSignedResponse(url, signed.headers.map().mapValues { (_, v) -> v.toList() })
        val json = JacksonUtils.toJson(value)
        val bytes = json.toByteArray(charset)

        val request = buildRequest(urlBuilder, HttpHeaders.empty(), r, body)
        return HttpResponse(request, 200, request.headers(), ByteArrayInputStream(bytes))
    }

    open fun pre(request: AwsS3PreRequest): AwsS3PreSignedResponse {
        val expire = request.expire
        checkNotNull(expire) { "pre sign expire must be not null!" }
        check(expire.isPositive) { "pre sign expire must be not positive!" }
        val response = call(request)
        val json = response.string()
        return JacksonUtils.toObj(json, AwsS3PreSignedResponse::class)
    }

}
